iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/aggregation/grouped/GroupedUserDefinedAggregateAccumulator.java

```diff
@@ -0,0 +1,139 @@
+import org.apache.tsfile.read.common.type.Type;
+import org.apache.tsfile.utils.Binary;
+import org.apache.tsfile.utils.RamUsageEstimator;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class GroupedUserDefinedAggregateAccumulator implements GroupedAccumulator {
+
+  private static final long INSTANCE_SIZE =
+      RamUsageEstimator.shallowSizeOfInstance(GroupedUserDefinedAggregateAccumulator.class);
+  private final AggregateFunction aggregateFunction;
+  private final ObjectBigArray<State> stateArray;
+  private final List<Type> inputDataTypes;
+
+  public GroupedUserDefinedAggregateAccumulator(
+      AggregateFunction aggregateFunction, List<Type> inputDataTypes) {
+    this.aggregateFunction = aggregateFunction;
+    this.stateArray = new ObjectBigArray<>();
+    this.inputDataTypes = inputDataTypes;
+  }
+
+  @Override
+  public long getEstimatedSize() {
+    return INSTANCE_SIZE;
+  }
+
+  @Override
+  public void setGroupCount(long groupCount) {
+    stateArray.ensureCapacity(groupCount);
+  }
+
+  private State getOrCreateState(int groupId) {
+    State state = stateArray.get(groupId);
+    if (state == null) {
+      state = aggregateFunction.createState();
+      stateArray.set(groupId, state);
+    }
+    return state;
+  }
+
+  @Override
+  public void addInput(int[] groupIds, Column[] arguments) {
+    RecordIterator iterator =
+        new RecordIterator(
+            Arrays.asList(arguments), inputDataTypes, arguments[0].getPositionCount());
+    int index = 0;
+    while (iterator.hasNext()) {
+      int groupId = groupIds[index++];
+      State state = getOrCreateState(groupId);
+      if (state == null) {
+        state = aggregateFunction.createState();
+        stateArray.set(groupId, state);
+      }
+      aggregateFunction.addInput(state, iterator.next());
+    }
+  }
+
+  @Override
+  public void addIntermediate(int[] groupIds, Column argument) {
+    checkArgument(
+        argument instanceof BinaryColumn
+            || (argument instanceof RunLengthEncodedColumn
+                && ((RunLengthEncodedColumn) argument).getValue() instanceof BinaryColumn),
+        "intermediate input and output of UDAF should be BinaryColumn");
+
+    for (int i = 0; i < groupIds.length; i++) {
+      if (!argument.isNull(i)) {
+        State otherState = aggregateFunction.createState();
+        Binary otherStateBinary = argument.getBinary(i);
+        otherState.deserialize(otherStateBinary.getValues());
+        aggregateFunction.combineState(getOrCreateState(groupIds[i]), otherState);
+      }
+    }
+  }
+
+  @Override
+  public void evaluateIntermediate(int groupId, ColumnBuilder columnBuilder) {
+    checkArgument(
+        columnBuilder instanceof BinaryColumnBuilder,
+        "intermediate input and output of UDAF should be BinaryColumn");
+    if (stateArray.get(groupId) == null) {
+      columnBuilder.writeBinary(new Binary(new byte[0]));
```

Comment:
when will this if happens? It seems that each group will always have a state? If so, better throw IllegalStateException here.

Reply:
done